app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,470 @@
+    }
+
+    /**
+     * Gets all elements and geometries inside [bbox]. This returns all nodes, all ways containing
+     * at least one of the nodes, and all relations containing at least one of the ways or nodes,
+     * and their geometries.
+     * If data is not cached, tiles containing the [bbox] are fetched from database and cached.
+     */
+    fun getMapDataWithGeometry(bbox: BoundingBox): MutableMapDataWithGeometry = synchronized(this) {
+        val requiredTiles = bbox.enclosingTilesRect(tileZoom).asTilePosSequence().toList()
+        val cachedTiles = spatialCache.getTiles()
+        val tilesToFetch = requiredTiles.filterNot { it in cachedTiles }
+        val tilesRectToFetch = tilesToFetch.minTileRect()
+
+        val result = MutableMapDataWithGeometry()
+        result.boundingBox = bbox
+        val nodes: Collection<Node>
+        if (tilesRectToFetch != null) {
+            // fetch needed data
+            val fetchBBox = tilesRectToFetch.asBoundingBox(tileZoom)
+            val (elements, geometries) = fetchMapData(fetchBBox)
+
+            // get nodes from spatial cache
+            // this may not contain all nodes, but tiles that were cached initially might
+            // get dropped when the caches are updated
+            // duplicate fetch might be unnecessary in many cases, but it's very fast anyway
+
+            nodes = HashSet<Node>(spatialCache.get(bbox))
+            update(updatedElements = elements, updatedGeometries = geometries, bbox = fetchBBox)
+
+            // return data if we need exactly what was just fetched
+            if (fetchBBox == bbox) {
+                val nodeGeometryEntries = elements.filterIsInstance<Node>().map { it.toElementGeometryEntry() }
+                result.putAll(elements, geometries + nodeGeometryEntries)
+                return result
+            }
+
+            // get nodes again, this contains the newly added nodes, but maybe not the old ones if cache was trimmed
+            nodes.addAll(spatialCache.get(bbox))
+        } else {
+            nodes = spatialCache.get(bbox)
+        }
+
+        val wayIds = HashSet<Long>(nodes.size / 5)
+        val relationIds = HashSet<Long>(nodes.size / 10)
+        for (node in nodes) {
+            wayIdsByNodeIdCache[node.id]?.let { wayIds.addAll(it) }
+            relationIdsByElementKeyCache[ElementKey(ElementType.NODE, node.id)]?.let { relationIds.addAll(it) }
+            result.put(node, ElementPointGeometry(node.position))
+        }
+        for (wayId in wayIds) {
+            result.put(wayCache[wayId]!!, wayGeometryCache[wayId])
+            relationIdsByElementKeyCache[ElementKey(ElementType.WAY, wayId)]?.let { relationIds.addAll(it) }
+        }
+        for (relationId in relationIds) {
+            result.put(relationCache[relationId]!!, relationGeometryCache[relationId])
+            // don't add relations of relations, because elementDao.getAll(bbox) also isn't doing that
+        }
```

Comment:
Does this actually need to happen if all is in cache already?
Or, does it need to happen for the portion of the data that is in cache already?

(I mean the updating of the `___IdsBy___IdCache`s, not the putting together of the result)

If it needs to happen, a short explanatory comment would be nice

Reply:
Actually the caches are not updated here at all.
`relationIdsByElementKeyCache[ElementKey(ElementType.NODE, node.id)]?.let { relationIds.addAll(it) }` just reads the cache and adds relation ids to `relationIds`.

I already noticed you don't like this kind of `?,let {}` things... it could be changed to
```kotlin
val nodeIds = relationIdsByElementKeyCache[ElementKey(ElementType.NODE, node.id)]
if (nodeIds != null)
    relationIds.addAll(nodeIds)
```

Reply:
> relationIdsByElementKeyCache[ElementKey(ElementType.NODE, node.id)]?.let { relationIds.addAll(it) } just reads the cache and adds relation ids to relationIds.

Well for me, this falls under "updating the cache". So, if e.g. the way is already in cache, the `relationIdsByElementKeyCache` should already be current, no?

Reply:
I don't understand... `relationIds` is never put to cache anywhere?
It's only used in `for (relationId in relationIds)`.

Reply:
Oh... yes. I thought `relationIds` referred to some map in the `relationIdsByElementKeyCache` or something. I guess I was confused because of the `let`, I'll change it to an `if` statement

Reply:
Or actually, I don't think changing it to if ... != null makes is any clearer, I guess I just didn't look properly. This is a case where `?.let` is perfectly fine, IMO.